File: GetInstance/Cdr.cs
using System.Xml.Linq;

namespace GetInstance
{
	public class Cdr
	{
		public GetInstanceResponse ExtractGetResponse(XDocument responseBody)
		{
			if (responseBody?.Element(XName.Get("GetInstanceDataResponse", "http://ffiec.gov/cdr/services/"))
				?.Element(XName.Get("GetInstanceDataResult", "http://ffiec.gov/cdr/services/")) is XElement getInstanceDataResult)
			{
				//string cdrText= String.Empty;
				//if (getInstanceDataResult.Value.Contains("&lt;"))
				{
					XElement cdr = getInstanceDataResult.Element(XName.Get("CdrServiceGetInstanceData", "http://Cdr.Business.Workflow.Schemas.CdrServiceGetInstanceData"));

					if (cdr == null)
					{
						//remove the declaration for xbrl document. another option would be processing xbrl under <InstanceDocument> as string and convert it to XDocument separately
						string cdrText = getInstanceDataResult.Value.Replace("&lt;", "<").Replace("&gt;", ">").Replace(@"<?xml version=""1.0"" encoding=""utf-8""?>", string.Empty);
						using (StringReader sr = new(cdrText))
							cdr = XElement.Load(new StringReader(cdrText));
					}

					if (cdr?.Name == XName.Get("CdrServiceGetInstanceData", "http://Cdr.Business.Workflow.Schemas.CdrServiceGetInstanceData") &&
						cdr.Element(XName.Get("Outputs")) is XElement outputs)
					{
						XElement returnCodeElement = outputs.Element("ReturnCode");
						XElement errorCodeElement = outputs.Element("ErrorCode");
						string codeText = (returnCodeElement ?? errorCodeElement)?.Value;
						XElement returnMessageElement = outputs.Element("ReturnMessage");
						XElement errorMessageElement = outputs.Element("ErrorMessage");
						string message = (returnMessageElement ?? errorMessageElement)?.Value;
						Console.WriteLine("Code: " + codeText);
						Console.WriteLine("Message: " + message);
						int? retcode = int.TryParse(codeText, out int code) ? code : null;
						XElement instanceDocuments = outputs.Element(XName.Get("InstanceDocuments"));
						return new GetInstanceResponse
						{
							Code = code,
							Message = message,
							InstanceDocuments = new XDocument(instanceDocuments)
						};
					}
				}
			}
			return null;
		}
	}
}

File: VB/SharedProject1/Cdr.vb
Imports System.IO
Imports System.Xml.Linq

Namespace GetInstance
	Public Class Cdr
		Private displayer As IDisplayer

		Public Sub New(ByVal displayer As IDisplayer)
			Me.displayer = displayer
		End Sub

		Public Function ExtractGetResponse(ByVal responseBody As XDocument) As GetInstanceResponse
			Dim code As Integer = Nothing, getInstanceDataResult As XElement = Nothing, outputs As XElement = Nothing

			If CSharpImpl.__Assign(getInstanceDataResult, TryCast(responseBody?.Element(XName.[Get]("GetInstanceDataResponse", "http://ffiec.gov/cdr/services/"))?.Element(XName.[Get]("GetInstanceDataResult", "http://ffiec.gov/cdr/services/")), XElement)) IsNot Nothing Then

				If True Then
					Dim cdr As XElement = getInstanceDataResult.Element(XName.[Get]("CdrServiceGetInstanceData", "http://Cdr.Business.Workflow.Schemas.CdrServiceGetInstanceData"))

					If cdr Is Nothing Then
						Dim cdrText As String = getInstanceDataResult.Value.Replace("&lt;", "<").Replace("&gt;", ">").Replace("<?xml version=""1.0"" encoding=""utf-8""?>", String.Empty)
						cdr = XElement.Load(New StringReader(cdrText))
					End If

					If cdr?.Name = XName.[Get]("CdrServiceGetInstanceData", "http://Cdr.Business.Workflow.Schemas.CdrServiceGetInstanceData") AndAlso CSharpImpl.__Assign(outputs, TryCast(cdr.Element(XName.[Get]("Outputs")), XElement)) IsNot Nothing Then
						Dim returnCodeElement As XElement = outputs.Element("ReturnCode")
						Dim errorCodeElement As XElement = outputs.Element("ErrorCode")
						Dim codeText As String = (If(returnCodeElement, errorCodeElement))?.Value
						Dim returnMessageElement As XElement = outputs.Element("ReturnMessage")
						Dim errorMessageElement As XElement = outputs.Element("ErrorMessage")
						Dim message As String = (If(returnMessageElement, errorMessageElement))?.Value
						Console.WriteLine("Code: " & codeText)
						Console.WriteLine("Message: " & message)
						Dim retcode As Integer? = If(Integer.TryParse(codeText, code), code, Nothing)
						Dim instanceDocuments As XElement = outputs.Element(XName.[Get]("InstanceDocuments"))
						Return New GetInstanceResponse With {
										.Code = code,
										.Message = message,
										.InstanceDocuments = New XDocument(instanceDocuments)
						}
					End If
				End If
			End If

			Return Nothing
		End Function

		Private Class CSharpImpl
			<Obsolete("Please refactor calling code to use normal Visual Basic assignment")>
			Shared Function __Assign(Of T)(ByRef target As T, value As T) As T
				target = value
				Return value
			End Function
		End Class
	End Class
End Namespace
